charts/sourcegraph/templates/_helpers.tpl

```diff
@@ -58,3 +58,42 @@ useGlobalTagAsDefault configuration
 
 {{- $top.Values.sourcegraph.image.repository }}/{{ $imageName }}:{{ default $defaultTag (index $top.Values $service "image" "tag") }}
 {{- end }}
+
+{{- define "sourcegraph.nodeSelector" -}}
+{{- $top := index . 0 }}
+{{- $service := index . 1 }}
+{{- $globalNodeSelector := (index $top.Values "sourcegraph" "nodeSelector") }}
+{{- $serviceNodeSelector := (index $top.Values $service "nodeSelector") }}
+nodeSelector:
+{{- if $serviceNodeSelector }}
+{{- $serviceNodeSelector | toYaml | trim | nindent 2 }}
+{{- else if $globalNodeSelector }}
```

Comment:
I was on the fence about whether this "if service, else if global" pattern was the right choice. It could mean a lot of duplication if there's a globally shared setting + service-specific settings. But I can't think of a nice way to support both situations (have the option to include global or exclude global, in addition to service-specific).

This scenario is pretty similar to the problem faced with the image tags, and I'm not very happy about the solution for that.

Your approach here means it is possible to get the right outcome for each service, and that's what's most important!

Reply:
> I was on the fence about whether this "if service, else if global" pattern was the right choice. It could mean a lot of duplication if there's a globally shared setting + service-specific settings. But I can't think of a nice way to support both situations (have the option to include global or exclude global, in addition to service-specific).
> 
> This scenario is pretty similar to the problem faced with the image tags, and I'm not very happy about the solution for that.
> 
> Your approach here means it is possible to get the right outcome for each service, and that's what's most important!

Are you talking about something like

```yaml
global:
  nodeSelector:
    zone: zone-1

frontend:
  nodeSelector:
    zone: zone-1
    node-purpose: generic
```

the users have to duplicate `zone: zone-1` in the per-service override, but we want to avoid the duplicatoin

Maybe we can merge global and per-service nodeSelector? so users can just do

```yaml
global:
  nodeSelector:
    zone: zone-1

frontend:
  nodeSelector:
    node-purpose: generic
```

The rendered `frontend-deployment` would have both `zone=zone-1` and `node-purpose: generic` in the `nodeSelector`?

This could potentially be put behind a flag, such as `global.mergeNodeSelector=true`?

tbh, I am not sure if this is a good idea. It's possible to users accidentally enable this flag and keep scratching their head wondering what's going on.

Reply:
> educating user about alias and anchor in yaml

Great point! I always forget about that feature (every time I wanted to use it, there were reasons it didn't work out). That's much better than trying to build in any fancy functionality.

Reply:
> > educating user about alias and anchor in yaml
> 
> Great point! I always forget about that feature (every time I wanted to use it, there were reasons it didn't work out). That's much better than trying to build in any fancy functionality.

LOL, I am never a fan of alias and anchor anyway. Also, the fact you can't merge sequences/list/array with it also feels off to me.

This would be a great candidate for docs improvement!